### PR TITLE
Release usb_camera_driver (ros2_usb_camera)

### PR DESCRIPTION
**Pull request by klintan**
First release of usb_camera_driver package. Please be gentle in your feedback as it's the first release of a package I've ever done! But excited to get it right, so let me know if there is anything I need to do or change.

Many thanks!

### CODE REVIEW

**Comment by klintan** [comment]
Hm I guess the version needs to be branch ? Fixed now, so if someone could retrigger that should be enough, if not I'll push something and revert back again to trigger.

File: foxy/distribution.yaml [hunk 1]
```diff
@@ -3237,6 +3237,23 @@ repositories:
       url: https://gitlab.com/ros-tracing/ros2_tracing.git
       version: foxy
     status: developed
+  ros2_usb_camera:
+    doc:
+      type: git
+      url: https://github.com/klintan/ros2_usb_camera.git
+      version: 0.1.0
```
Comment (jacobperron) [inline foxy/distribution.yaml:3244]: The `doc` and `source` entries should point to a branch, not a tag.

**Review by jacobperron — COMMENTED** [review]
Typically, rosdistro pull requests are opened automatically with bloom. Was there an issue/reason to not let bloom handle it?

File: foxy/distribution.yaml (hunk 1 above)
Comment (klintan) [reply foxy/distribution.yaml:3244]: it's a branch now, do I need to remove the tag as well?

**Comment by klintan** [comment]
> Typically, rosdistro pull requests are opened automatically with bloom. Was there an issue/reason to not let bloom handle it?

Yes it unfortunately it failed at the very last step for the PR and just recommended to do a manual PR.

**Comment by jacobperron** [comment]
> Yes it unfortunately it failed at the very last step for the PR and just recommended to do a manual PR.

I'm not sure exactly what failed, but if it was an authorization issue, I recommend taking a look at http://wiki.ros.org/bloom/Tutorials/GithubManualAuthorization

File: foxy/distribution.yaml (hunk 1 above)
Comment (jacobperron) [reply foxy/distribution.yaml:3244]: I think we should avoid branch names that are the same as tag names. I recommend creating a tag (`0.1.1`) at a point in history on the development branch (`foxy-devel`). Then, the `source` and `doc` entry would be `foxy-devel`, and then re-run bloom updating the development branch to `foxy-devel`. E.g. 

```
bloom-release --rosdistro foxy --track foxy ros2_usb_camera --edit
```

**Review by jacobperron — COMMENTED** [review]
As a separate concern, I think the prefix `ros2_` of the package name is redundant and could be removed (see naming guidelines in [REP 144](https://ros.org/reps/rep-0144.html). I suggest renaming the package to `usb_camera`.

---

It's good that you you reference the existing **libuvc** and **usb_cam** packages in your README, but it begs the question, why not just use one of them (I.e. port to ROS 2)?
It looks like there has been some work done on porting **usb_cam**, and there's interest in helping maintain it (https://github.com/ros-drivers/usb_cam/issues/135).

I'm not trying to undermine the work you've done, but it would be good to document how this driver is different (and why someone might use it over the existing drivers). Maybe efforts could be combined?

**Comment by hidmic** [comment]
@klintan friendly ping.

**Comment by klintan** [comment]
I'll get it done this weekend :)

**Comment by klintan** [comment]
> As a separate concern, I think the prefix `ros2_` of the package name is redundant and could be removed (see naming guidelines in [REP 144](https://ros.org/reps/rep-0144.html). I suggest renaming the package to `usb_camera`.
> 
> It's good that you you reference the existing **libuvc** and **usb_cam** packages in your README, but it begs the question, why not just use one of them (I.e. port to ROS 2)?
> It looks like there has been some work done on porting **usb_cam**, and there's interest in helping maintain it ([ros-drivers/usb_cam#135](https://github.com/ros-drivers/usb_cam/issues/135)).
> 
> I'm not trying to undermine the work you've done, but it would be good to document how this driver is different (and why someone might use it over the existing drivers). Maybe efforts could be combined?

The package name in the `package.xml` is `usb_camera_driver`, however the name from the bloom process seems to pick up the repository name? Is the track name that what becomes the name in `rosdistro` ? so I need to run 
`bloom-release --rosdistro foxy --track foxy usb_camera --edit` or ``bloom-release --rosdistro foxy --track foxy usb_camera_driver --edit` for that name?

I added some references in the readme for alternate packages and why someone might use this instead. I couldn't find the Ros2 version of `libuvc` though, could you provide a link to the ros2 version and I'lll add it to the readme as well.

Thanks for your patience.

**Comment by jacobperron** [comment]
As far as I know, the track name has nothing to do with rosdistro; it's meant for selecting different release configurations (you probably don't need to worry about this). Note, the last argument before `--edit` is the package name. I'm not sure if it has to match the repo, so you can probably run

```
bloom-release --rosdistro foxy --track foxy usb_camera --edit
```

and generate a new rosdistro PR. You should probably create a new release repository to remove the branches created from the first release. E.g. delete and re-create https://github.com/klintan/ros2_usb_camera-release

---

I'm not aware of a ROS 2 version of libuvc. It was more a question of why a new package vs. porting the existing one to ROS 2.

**Comment by jacobperron** [comment]
@klintan Friendly ping :slightly_smiling_face:

**Comment by klintan** [comment]
I'll try again tonight, third times a charm :)

**Comment by tfoote** [comment]
There's an active effort to port the existing camera driver: https://discourse.ros.org/t/usb-cam-orphaned-offer-to-support/19367/8

It would be great to merge those efforts. 

The entry you're seeing in this diff comes from the repository name. It is not actually used anywhere else. You can just change it here in the diff. And if you do that such that the package name is the same as the repository name it won't need to list them.

The other change is that the branches `foxy-devel` need to be substituted into the version lines for the doc and source entries. But all the changes needed are only in this PR content. 

Related there's the ongoing effort to move usb_cam forward. It would be great to merge this effort with that one so we can have one clear one for the community to focus their efforts on. https://discourse.ros.org/t/usb-cam-orphaned-offer-to-support/19367/8

Without an update I plan to close this next week. It can always be reopened or a replacement ticket opened when the above is resolved.

**Comment by klintan** [comment]
> There's an active effort to port the existing camera driver: https://discourse.ros.org/t/usb-cam-orphaned-offer-to-support/19367/8
> 
> It would be great to merge those efforts.
> 
> The entry you're seeing in this diff comes from the repository name. It is not actually used anywhere else. You can just change it here in the diff. And if you do that such that the package name is the same as the repository name it won't need to list them.
> 
> The other change is that the branches `foxy-devel` need to be substituted into the version lines for the doc and source entries. But all the changes needed are only in this PR content.
> 
> Related there's the ongoing effort to move usb_cam forward. It would be great to merge this effort with that one so we can have one clear one for the community to focus their efforts on. https://discourse.ros.org/t/usb-cam-orphaned-offer-to-support/19367/8
> 
> Without an update I plan to close this next week. It can always be reopened or a replacement ticket opened when the above is resolved.

Thanks for the pointers, and I def agree that several camera versions are maybe not great. However I was trying and hoping that this package could be a bit simpler and based on OpenCV, I think looking at the usb_cam library was a bit more complex and had no OpenCV dependency.

Sorry this has taken some time, a lot of other things that has come up. But I will make another try thursday next week. Maybe this weekend if I get the time :) 

As you said, worst case I'll create a new release ticket.

**Comment by nuclearsandwich** [comment]
@klintan thanks for contributing and iterating thus far! I am going to close this PR based on the conversation above but wanted to add that you can re-run `bloom-release` with the `--pull-request-only` flag and it will attempt to create the pull request for the latest release on the specified track instead of creating a new release, which can be very helpful if the pull request creation fails for some reason.

**Pull request closed without merging**